toolbox/pkg_check/main.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"golang.org/x/net/context"
+)
+
+const (
+	jobName = "JOB_NAME"
+	buildID = "BUILD_ID"
+)
+
+var (
+	reportFile      = flag.String("report_file", "codecov.report", "Package code coverage report.")
+	requirementFile = flag.String("requirement_file", "codecov.requirement", "Package code coverage requirement.")
+	gcsBucket       = flag.String("bucket", "istio-code-coverage", "gcs bucket")
+)
+
+type codecovChecker struct {
+	codeCoverage  map[string]float64
+	report        string
+	requirement   string
+	failedPackage []string
+	bucket        string
+}
+
+func (c *codecovChecker) parseReport() error {
+	f, err := os.Open(c.report)
+	if err != nil {
+		log.Printf("Failed to open report file %s", c.report)
+		return err
+	}
+	defer func() {
+		if err = f.Close(); err != nil {
+			log.Printf("Failed to close file %s, %v", c.report, err)
+		}
+	}()
+
+	scanner := bufio.NewScanner(f)
+
+	//Report example: "ok  	istio.io/mixer/adapter/denyChecker	0.023s	coverage: 100.0% of statements"
+	//Report example: "?   istio.io/mixer/adapter/denyChecker/config	[no test files]"
+	//expected output: c.codeCoverage["istio.io/mixer/adapter/denyChecker"] = 100
+	for scanner.Scan() {
```

Comment:
This seems very flaky. You should probably use a regular expression to extract the information.